From 55ca1d35eea4e269e455835b48e3aabb342d7855 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 18 Apr 2025 14:29:33 +0800
Subject: [PATCH 190/236] mtk: hostapd: prevent NULL pointer access in
 hostapd_free_hapd_data()

In the hostapd_free_hapd_data(), other hapd on the same
hostapd_iface might already be freed so that iface->bss[j] is NULL.

For example, hostapd_setup_bss() fails on the first hapd and returns
the failure to hostapd_add_iface(). In the error handling of
hostapd_add_iface(), it goes through each hapd on the iface and does
the deinitialization, which including calling hostapd_free_hapd_data()
and setting iface->bss[i] to NULL. In other word, iface->bss[1] enters
it deinitialization with iface->bss[0] being NULL, and a NULL pointer
access happens in hostapd_free_hapd_data().

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/hostapd.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 5cba10899..8f7fa8ed7 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -548,10 +548,13 @@ void hostapd_free_hapd_data(struct hostapd_data *hapd)
 			struct hostapd_iface *iface = ifaces->iface[i];
 			size_t j;
 
+			if (hapd->iface == iface)
+				continue;
+
 			for (j = 0; iface && j < iface->num_bss; j++) {
 				struct hostapd_data *h = iface->bss[j];
 
-				if (hapd == h)
+				if (!h)
 					continue;
 				if (h->radius == hapd->radius)
 					h->radius = NULL;
-- 
2.45.2

